src/logic/add.rs

```diff
@@ -0,0 +1,265 @@
+                            //if the ApInt was relatively random this should rarely happen
+                            x[i] = v;
+                        }
+                    }
+                }
+            }
+        }
+        self.clear_unused_bits();
+    }
+
+    /// Increments this `ApInt` by one and returns the result.
+    pub fn into_wrapping_inc(self) -> ApInt {
+        forward_mut_impl(self, ApInt::wrapping_inc)
+    }
+
+    /// Decrements this `ApInt` by one inplace.
+    pub fn wrapping_dec(&mut self) {
+        match self.access_data_mut() {
+            DataAccessMut::Inl(x) => {
+                *x = x.wrapping_sub(Digit::one());
+            }
+            DataAccessMut::Ext(x) => {
+                for i in 0..x.len() {
+                    match x[i].overflowing_sub(Digit::one()) {
+                        (v,false) => {
+                            x[i] = v;
+                            break;
+                        }
+                        (v,true) => {
+                            //if the ApInt was relatively random this should rarely happen
+                            x[i] = v;
+                        }
+                    }
+                }
+            }
+        }
+        self.clear_unused_bits();
+    }
+
+    /// Decrements this `ApInt` by one and returns the result.
+    pub fn into_wrapping_dec(self) -> ApInt {
+        forward_mut_impl(self, ApInt::wrapping_dec)
+    }
+
+    /// Negates this `ApInt` inplace.
+    pub fn wrapping_neg(&mut self) {
+        self.bitnot();
+        self.wrapping_inc();
+        //`wrapping_inc` handles clearing the unused bits
+    }
+
+    /// Negates this `ApInt` and returns the result.
+    pub fn into_wrapping_neg(self) -> ApInt {
+        forward_mut_impl(self, ApInt::wrapping_neg)
+    }
+
+    /// Add-assigns `rhs` to `self` inplace.
+    pub fn wrapping_add_assign(&mut self, rhs: &ApInt) -> Result<()> {
+        match self.zip_access_data_mut_self(rhs)? {
+            Inl(lhs, rhs) => {
+                *lhs = lhs.wrapping_add(rhs);
+            }
+            Ext(lhs, rhs) => {
+                let (temp, mut carry) = lhs[0].carrying_add(rhs[0]);
+                lhs[0] = temp;
+                for i in 1..lhs.len() {
+                    let temp = lhs[i].dd()
+                        .wrapping_add(rhs[i].dd())
+                        .wrapping_add(carry.dd());
+                    lhs[i] = temp.lo();
+                    carry = temp.hi();
+                }
+            }
+        }
+        self.clear_unused_bits();
+        Ok(())
+    }
+
+    /// Adds `rhs` to `self` and returns the result.
+    pub fn into_wrapping_add(self, rhs: &ApInt) -> Result<ApInt> {
+        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_add_assign)
+    }
+
+    /// Add-assigns `rhs` to `self` inplace, and returns a boolean indicating if overflow occured,
+    /// according to the **unsigned** interpretation of overflow.
+    // TODO: add tests
```

Comment:
Do we already have tests for this?

Reply:
Just now, I wrote proper tests for this function, and guess what I found.
Another bug involving odd width `ApInt`s! I misunderstood the `excess_bit`s function (in fact there is a note you put there about changing the name of the function). The fix was one line and the fuzz tester which was the only thing depending on it did not have any hidden bugs for me fortunately.